### PR TITLE
Simplify Plugin Manager for official plugins

### PR DESCRIPTION
**Pull request by dadoonet**
Plugin Manager can now use another simplified form when a user wants to install an official plugin hosted at elasticsearch download service.

The form we use is:

``` sh
bin/plugin install pluginname
```

As plugins share now the same version as elasticsearch, we can automatically guess what is the exact current version of the plugin manager script.

Also, download service will now use `/org.elasticsearch.plugins/pluginName/pluginName-version.zip` URL path to download a plugin.

If the older form is provided (`user/plugin/version` or `user/plugin`), we will still use:
- elasticsearch download service at `/user/plugin/plugin-version.zip`
- maven central with groupIp=user, artifactId=plugin and version=version
- github with user=user, repoName=plugin and tag=version
- github with user=user, repoName=plugin and branch=master if no version is set

Note that community plugin providers can use other download services by using `--url` option.

If you try to use the new form with a non core elasticsearch plugin, the plugin manager will reject
it and will give you all known core plugins.

```
Usage:
    -u, --url     [plugin location]   : Set exact URL to download the plugin from
    -i, --install [plugin name]       : Downloads and installs listed plugins [*]
    -t, --timeout [duration]          : Timeout setting: 30s, 1m, 1h... (infinite by default)
    -r, --remove  [plugin name]       : Removes listed plugins
    -l, --list                        : List installed plugins
    -v, --verbose                     : Prints verbose messages
    -s, --silent                      : Run in silent mode
    -h, --help                        : Prints this help message

 [*] Plugin name could be:
     elasticsearch-plugin-name    for Elasticsearch 2.0 Core plugin (download from download.elastic.co)
     elasticsearch/plugin/version for elasticsearch commercial plugins (download from download.elastic.co)
     groupId/artifactId/version   for community plugins (download from maven central or oss sonatype)
     username/repository          for site plugins (download from github master)

Elasticsearch Core plugins:
 - elasticsearch-analysis-icu
 - elasticsearch-analysis-kuromoji
 - elasticsearch-analysis-phonetic
 - elasticsearch-analysis-smartcn
 - elasticsearch-analysis-stempel
 - elasticsearch-cloud-aws
 - elasticsearch-cloud-azure
 - elasticsearch-cloud-gce
 - elasticsearch-delete-by-query
 - elasticsearch-lang-javascript
 - elasticsearch-lang-python
```

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 1]
```diff
@@ -582,17 +582,24 @@ public static PrintStream getOut() {
...
             if (version != null) {
-                // Elasticsearch download service
-                addUrl(urls, "http://download.elasticsearch.org/" + user + "/" + repo + "/" + repo + "-" + version + ".zip");
-                // Maven central repository
-                addUrl(urls, "http://search.maven.org/remotecontent?filepath=" + user.replace('.', '/') + "/" + repo + "/" + version + "/" + repo + "-" + version + ".zip");
-                // Sonatype repository
-                addUrl(urls, "https://oss.sonatype.org/service/local/repositories/releases/content/" + user.replace('.', '/') + "/" + repo + "/" + version + "/" + repo + "-" + version + ".zip");
-                // Github repository
-                addUrl(urls, "https://github.com/" + user + "/" + repo + "/archive/" + version + ".zip");
+                // Elasticsearch new download service uses groupId org.elastcisearch.plugins from 2.0.0
```
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: Argh! Typo :)

**Comment by dadoonet** [comment]
@tlrx Could you review it please?

File: docs/reference/modules/plugins.asciidoc [hunk 2]
```diff
@@ -14,19 +14,26 @@ and more.
...
+`plugins` directory, or using the `plugin` script.
 
 Installing plugins typically take the following form:
 
 [source,shell]
 -----------------------------------
+bin/plugin --install plugin_name
+-----------------------------------
+
+The plugin will be automatically downloaded in this case from `download.elastic.co` download service using the
```
Comment (spinscale) [inline docs/reference/modules/plugins.asciidoc]: you are using `download.elastic.co` here and `download.elasticsearch.org` in the code. Intended

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 3]
```diff
@@ -638,6 +645,13 @@ static PluginHandle parse(String name) {
                 }
             }
 
+            if (version == null && user == null) {
+                // This is the new way with es 2.0.0 to download official plugins
+                // We use bin/plugin install pluginname
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: you could remove a fair share of this comment by just having a boolean

```java
boolean isOfficial20Plugin = version == null && user == null && !Strings.empty(repo);
```

File: docs/reference/modules/plugins.asciidoc (hunk 2 above)
Comment (dadoonet) [inline docs/reference/modules/plugins.asciidoc]: Well... I did not try to change the old behavior but you're right. I guess download.elastic.co works as well as download.elasticsearch.org? Need to test it and if it works, I will update the PR accordingly.

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java (hunk 3 above)
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: Nice. Will do.

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 4]
```diff
@@ -582,17 +582,24 @@ public static PrintStream getOut() {
...
+                    addUrl(urls, "http://download.elasticsearch.org/org.elasticsearch.plugins/" + repo + "/" + repo + "-" + version + ".zip");
+                } else {
+                    // Elasticsearch old download service
+                    addUrl(urls, "http://download.elasticsearch.org/" + user + "/" + repo + "/" + repo + "-" + version + ".zip");
+                    // Maven central repository
+                    addUrl(urls, "http://search.maven.org/remotecontent?filepath=" + user.replace('.', '/') + "/" + repo + "/" + version + "/" + repo + "-" + version + ".zip");
+                    // Sonatype repository
+                    addUrl(urls, "https://oss.sonatype.org/service/local/repositories/releases/content/" + user.replace('.', '/') + "/" + repo + "/" + version + "/" + repo + "-" + version + ".zip");
+                    // Github repository
+                    addUrl(urls, "https://github.com/" + user + "/" + repo + "/archive/" + version + ".zip");
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: not really scope of this PR, but otherwise I will do it.. if we use `String.format()` for those URLs, they will be more readable

Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: Totally agreed! Will update.

File: docs/reference/modules/plugins.asciidoc (hunk 2 above)
Comment (spinscale) [inline docs/reference/modules/plugins.asciidoc]: lets especially ensure none of those goes away :-)

**Comment by spinscale** [comment]
one last thing: do we need to retain bwc compatibility or at least help the user in case an older plugin tries to be installed.

Checking the official [kuromoji docs](https://github.com/elastic/elasticsearch-analysis-kuromoji) this is the way to install the plugin

```bash
bin/plugin install elasticsearch/elasticsearch-analysis-kuromoji/2.6.0
```

what happens if I do this with master? Will it work? Will it break? Should be catch this in the PluginManager and write it to `bin/plugin install kuromoji` or just output a big warning, that the user tries to install something old? Do we need documentation for this, that official plugins are now installed differently?

I havent formed an opinion here yet (except that we have to document it, not sure, if we should try to be smart or just issue a warning), but we should get a consensus

**Comment by clintongormley** [comment]
> Checking the official kuromoji docs this is the way to install the plugin

We're moving the plugin documentation for 2.0, but we will need to preserve the old instructions for the 1.x branches.

**Comment by spinscale** [comment]
@clintongormley sure thing. I meant, people might have put that in some automated installation scripts. And it is not given that they just go to the website they are used to go to, to extract the installation information and copy paste this into a terminal

I just played around with the current impl. Installing kuromoji without any version specified results in a cryptic error message on 1.6

```
bin/plugin install elasticsearch/elasticsearch-analysis-kuromoji
-> Installing elasticsearch/elasticsearch-analysis-kuromoji...
Trying https://github.com/elasticsearch/elasticsearch-analysis-kuromoji/archive/master.zip...
Downloading ..............................DONE
Installed elasticsearch/elasticsearch-analysis-kuromoji into /tmp/elasticsearch-1.6.0/plugins/analysis-kuromoji
Usage:
    -u, --url     [plugin location]   : Set exact URL to download the plugin from
    -i, --install [plugin name]       : Downloads and installs listed plugins [*]
    -t, --timeout [duration]          : Timeout setting: 30s, 1m, 1h... (infinite by default)
    -r, --remove  [plugin name]       : Removes listed plugins
    -l, --list                        : List installed plugins
    -v, --verbose                     : Prints verbose messages
    -s, --silent                      : Run in silent mode
    -h, --help                        : Prints this help message

 [*] Plugin name could be:
     elasticsearch/plugin/version for official elasticsearch plugins (download from download.elasticsearch.org)
     groupId/artifactId/version   for community plugins (download from maven central or oss sonatype)
     username/repository          for site plugins (download from github master)

Message:
   Error while installing plugin, reason: IllegalArgumentException: Plugin installation assumed to be site plugin, but contains source code, aborting installation.
```

My proposal is to spit out an error message for all the new official plugins 2.0 if someone tries to install kuromoji and all the others the old way, telling the user this is now an official supported plugin to be installed via `bin/plugin install kuromoji` in case the user typed `bin/plugin install elasticsearch/elasticsearch-analysis-kuromoji[/$VERSION]`

File: docs/reference/modules/plugins.asciidoc (hunk 2 above)
Comment (dadoonet) [inline docs/reference/modules/plugins.asciidoc]: So elastic.co url works fine. BTW We were using http instead of https. Do you think I should use https from now?

**Comment by clintongormley** [comment]
++

File: docs/reference/modules/plugins.asciidoc (hunk 2 above)
Comment (spinscale) [inline docs/reference/modules/plugins.asciidoc]: can we do this in another PR maybe (dont wanna bloat this one and make sure it gets in). Lets keep it as is for now, but I am all for https with a possibility to fallback to http (via CLI parameter).

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 5]
```diff
@@ -582,17 +583,24 @@ public static PrintStream getOut() {
...
-                addUrl(urls, "http://download.elasticsearch.org/" + user + "/" + repo + "/" + repo + "-" + version + ".zip");
-                // Maven central repository
-                addUrl(urls, "http://search.maven.org/remotecontent?filepath=" + user.replace('.', '/') + "/" + repo + "/" + version + "/" + repo + "-" + version + ".zip");
-                // Sonatype repository
-                addUrl(urls, "https://oss.sonatype.org/service/local/repositories/releases/content/" + user.replace('.', '/') + "/" + repo + "/" + version + "/" + repo + "-" + version + ".zip");
-                // Github repository
-                addUrl(urls, "https://github.com/" + user + "/" + repo + "/archive/" + version + ".zip");
+                // Elasticsearch new download service uses groupId org.elasticsearch.plugins from 2.0.0
+                if (user == null) {
+                    addUrl(urls, String.format("https://download.elastic.co/org.elasticsearch.plugins/%1$s/%1$s-%2$s.zip", repo, version));
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: have you run this through maven? I think `String.format()` without a locale is part of the forbidden API

**Comment by dadoonet** [comment]
@spinscale Pushed new changes.

Was thinking of detecting when user ask for `bin/plugin install elasticsearch/foo/bar`. We can't really reject such a request because of commercial plugins which don't use (yet?) `org.elasticsearch.plugins`.

Or do you want us to have a static builtin list of plugins which moved to elasticsearch repo?

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java (hunk 5 above)
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: Very true! Fixing

Comment (spinscale) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: shouldnt this solve the problem and not the above

```
String.format(Locale.ROOT, "https://download.elastic.co/org.elasticsearch.plugins/%s/%s-%s.zip", repo, version));
```

**Comment by spinscale** [comment]
Hm, the best thing would be, if we could find out during build/packaging time which plugins we ship and only act on those. This would allow us to remove/add plugins arbitrarily per release. Do you think that is feasible?

**Comment by dadoonet** [comment]
I don't think it is and I'm not sure we should. It's like having a circular dependency between plugin and core IMO.

The "easiest" workaround would be to publish our commercial plugins (and change the groupId to make all that consistent) to `org.elasticsearch.commercial`... or something.

I would not try to fix that for now so we don't block this PR but this is something we can think about for GA. WDYT?

File: core/src/test/java/org/elasticsearch/plugins/PluginManagerUnitTests.java [hunk 6]
```diff
@@ -55,4 +56,14 @@ public void testThatConfigDirectoryCanBeOutsideOfElasticsearchHomeDirectory() th
...
         assertThat(configDirPath, is(expectedDirPath));
     }
+
+    @Test
+    public void testSimplifiedNaming() throws IOException {
+        String pluginName = randomAsciiOfLength(10);
+        PluginManager.PluginHandle handle = PluginManager.PluginHandle.parse(pluginName);
+        assertThat(handle.urls(), hasSize(1));
+        assertThat(handle.urls().get(0).getHost(), is("download.elasticsearch.org"));
+        assertThat(handle.urls().get(0).getPath(), is("/org.elasticsearch.plugins/" + pluginName + "/" +
```
Comment (dadoonet) [inline core/src/test/java/org/elasticsearch/plugins/PluginManagerUnitTests.java]: @spinscale Argh! This one is also forbidden `URL.getPath()` :)

**Comment by spinscale** [comment]
not sure about this, 

@clintongormley what do you think? If we hardcode such a check, we might have problems, moving such a plugin out of the core again. Or we just make sure the plugin list is in sync. And then we can leave it hardcoded. As I think this is a huge improvement for the not-knowing user, I am for the hardcoded solution for now, which means we need to pay attention when we add or remove plugins to the elasticsearch repo.

**Comment by clintongormley** [comment]
Besides Shield etc, there are also things like https://github.com/elastic/elasticsearch-migration.

I think hard coding is a simple solution. It's not like the list changes often, and we can add a test to https://github.com/elastic/elasticsearch/blob/master/dev-tools/smoke_test_plugins.py to ensure that the list is up to date.

**Comment by dadoonet** [comment]
@spinscale I added a new commit. Let me know

**Comment by spinscale** [comment]
should we also add a paragraph for the migration docs for this?

**Comment by dadoonet** [comment]
@spinscale I added 2 more commits

**Comment by spinscale** [comment]
we also need to list the plugins that are considered official somewhere I guess

**Comment by dadoonet** [comment]
@spinscale I added the plugin list in the migration guide. All the documentation will be updated with this PR #12040.

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 7]
```diff
@@ -595,7 +610,8 @@ private static void displayHelp(String message) {
         SysOut.println("    -h, --help                        : Prints this help message");
         SysOut.newline();
         SysOut.println(" [*] Plugin name could be:");
-        SysOut.println("     elasticsearch/plugin/version for official elasticsearch plugins (download from download.elasticsearch.org)");
+        SysOut.println("     plugin for official elasticsearch plugins (download from download.elastic.co for elasticsearch >= 2.0)");
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: maybe just write `elasticsearch-plugin-name     Elasticsearch 2.0 Core plugin` (fixes the awkward spacing problem)

Also the phrasing assumes that elasticsearch commercial plugins are not official somewhat? Not sure if its just me being non-native speaking.

**Comment by spinscale** [comment]
I think we are pretty much there, I will do some further command line testing and report back.

One last thing: If someone uses the `bin/plguin install elasticsearch-analysis-kuromoji` pattern and the plugin cannot be found, do you think it makes sense to just list the available plugins in the output to help the user?

**Comment by spinscale** [comment]
the log4j config seems to have a problem when using the zip package

```
bin/plugin install elasticsearch/license/latest
-> Installing elasticsearch/license/latest...
Trying http://download.elastic.co/elasticsearch/license/license-latest.zip...
Downloading ............................................DONE
log4j:WARN No appenders could be found for logger (org.elasticsearch.bootstrap).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Installed elasticsearch/license/latest into /tmp/elasticsearch-2.0.0-SNAPSHOT/plugins/license
```

**Comment by spinscale** [comment]
**Update**: This is a problem in master, not in this branch, I will take a look

**Comment by dadoonet** [comment]
@spinscale Yeah. See also #12064
I just hit the same issues in plugins.

**Comment by dadoonet** [comment]
@spinscale Added a new commit based on your comments. I just print now the list of Core plugins.
I could also replace this list by `bin/plugin install XXX` for each element... Not sure what is the best option.

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 8]
```diff
@@ -80,6 +80,21 @@
                     "plugin.bat",
                     "service.bat").build();
 
+    private static final ImmutableSet<Object> OFFICIAL = ImmutableSet.builder()
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: why no `private static final ImmutableSet<String> OFFICIAL_PLUGINS = ImmutableSet.<String>builder()`? any generics voodoo I miss? :-)

**Comment by spinscale** [comment]
I would only output the official plugin list, if the user has tried to install a plugin that does not exist (personal taste, freel free to ignore)... apart from that LGTM

**Comment by spinscale** [comment]
all right, rant tests with `-Dtests.slow` and `-Dtests.network`.. LGTM

**Comment by dadoonet** [comment]
Thank you Alex!

**Comment by dadoonet** [comment]
@spinscale I was reading again this PR and I think I wrote something wrong.
I was considering `org.elasticsearch.plugins` as the `groupId` but it should be `org.elasticsearch.plugin` (for example: https://github.com/elastic/elasticsearch/blob/master/plugins/analysis-icu/pom.xml#L8) 

I will come with an update.
